### PR TITLE
Implement state distribution in result set.

### LINKED ISSUE #738: Add state distributions to result set
Following #737 we can now add to the result sets (base and 'from file') consistently (before that they were being built in two separate ways: now code is not duplicated).

The match class and interaction utils also now have functions to calculate the state distribution for a match: counts of turns spent in each of `CC, CD, DC, DD`. 

It should be easy enough to add that to the result set. I suggest we have, similar to `results.cooperation` and `results.normalised_cooperation`:

```
results.state_distribution  # A list of lists (1 for each player);of lists (1 for each opponent); of 4 tuples of total (across all repetitions) of counts of `CC, CD, DC, DD`
results.normalised_state_distribution  # # A list of lists (1 for each player);of lists (1 for each opponent); of 4 tuples of total (across all repetitions) of normalised counts of `CC, CD, DC, DD`
```

I also suggest we add 4 more columns to the output of  `results.summarise`, which are the mean (across all opponents) normalised counts for each state.

### PR DESCRIPTION
**Pull request by drvinceknight**
Closes #738

Add state distribution and normalised state distribution to the result set.

Also adds the normalised state distribution to the summary.

### CODE REVIEW

File: axelrod/result_set.py [hunk 1]
```diff
@@ -326,6 +326,26 @@ def build_ranking(self):
         return sorted(range(self.nplayers),
                       key=lambda i: -nanmedian(self.normalised_scores[i]))
 
+    def build_normalised_state_distribution(self):
```
Comment (marcharper) [inline axelrod/result_set.py]: Needs an underscore? Seems like we're being inconsistent with the build* functions

File: axelrod/result_set.py [hunk 2]
```diff
@@ -326,6 +326,26 @@ def build_ranking(self):
...
+            Normalised state distribution. A list of lists of counter objects:
+
+            Dictionary where the keys are the states and the values are a
+            normalized counts of the number of times that state occurs.
+        """
+        norm = []
+        for player in self.state_distribution:
+            counters = []
+            for counter in player:
+                total = sum(counter.values(), 0.0)
```
Comment (marcharper) [inline axelrod/result_set.py]: Is the 0.0 necessary here? Is it possible for sum to be zero?

File: axelrod/result_set.py [hunk 3]
```diff
@@ -401,6 +423,13 @@ def _update_cooperation(self, p1, p2, cooperations):
         self.cooperation[p1][p2] += cooperations[0]
         self.cooperation[p2][p1] += cooperations[1]
 
+    def _update_state_distribution(self, p1, p2, counter):
+        self.state_distribution[p1][p2] += counter
```
Comment (marcharper) [inline axelrod/result_set.py]: Docstring?

File: axelrod/result_set.py [hunk 4]
```diff
@@ -401,6 +423,13 @@ def _update_cooperation(self, p1, p2, cooperations):
...
+
+        counter[('C', 'D')], counter[('D', 'C')] = (counter[('D', 'C')],
```
Comment (marcharper) [inline axelrod/result_set.py]: I like the swap but not the line break.... I guess there's no getting around that, but maybe it would be more clear if broken after the = sign

Comment (marcharper) [inline axelrod/result_set.py]: Could shorten with `C, D = actions.C, actions.D` at the top of file...

File: axelrod/result_set.py [hunk 5]
```diff
@@ -548,13 +580,28 @@ def summarise(self):
         median_wins = map(nanmedian, self.wins)
 
         self.player = namedtuple("Player", ["Rank", "Name", "Median_score",
-                                            "Cooperation_rating", "Wins"])
+                                            "Cooperation_rating", "Wins",
+                                            "CC_rate", "CD_rate", "DC_rate",
+                                            "DD_rate"])
+
+        states = [('C', 'C'), ('C', 'D'), ('D', 'C'), ('D', 'D')]
```
Comment (marcharper) [inline axelrod/result_set.py]: Could use `C, D = actions.C, actions.D`

File: axelrod/result_set.py [hunk 6]
```diff
@@ -548,13 +580,28 @@ def summarise(self):
...
+        state_prob = []
+        for i, player in enumerate(self.normalised_state_distribution):
+            counts = []
+            for state in states:
+                counts.append(sum([opp[state] for j, opp in enumerate(player)
```
Comment (marcharper) [inline axelrod/result_set.py]: Two lines here? One for the sum, one for the append so there's no line break

File: axelrod/result_set.py [hunk 7]
```diff
@@ -548,13 +580,28 @@ def summarise(self):
...
+                                   if i != j]))
+            try:
+                counts = [c / sum(counts) for c in counts]
```
Comment (marcharper) [inline axelrod/result_set.py]: Any change for an integer division error here?

File: axelrod/result_set.py [hunk 8]
```diff
@@ -548,13 +580,28 @@ def summarise(self):
...
+
+        summary_data = list(zip(self.players, median_scores,
+                                self.cooperating_rating, median_wins))
 
-        summary_data = [perf for perf in zip(self.players,
-                                             median_scores,
-                                             self.cooperating_rating,
-                                             median_wins)]
-        summary_data = [self.player(rank, *summary_data[i]) for
+        summary_data = [self.player(rank, *(list(summary_data[i])
```
Comment (marcharper) [inline axelrod/result_set.py]: This is a bit hard to follow... could use a readability upgrade?

File: docs/tutorials/getting_started/tournament_results.rst [hunk 9]
```diff
@@ -20,6 +20,10 @@ This tutorial will show you how to access the various results of a tournament:
 - Payoff difference means: the mean score differences.
 - Cooperation counts: the number of times each player cooperated.
 - Normalised cooperation: cooperation count per turn.
+- Normalised cooperation: cooperation count per turn.
+- State distribution: the count of each type of state of a match
```
Comment (marcharper) [inline docs/tutorials/getting_started/tournament_results.rst]: Might be good to explain what we mean by state distribution -- that it's two rounds of play rather than some other amount.

Comment (drvinceknight) [inline docs/tutorials/getting_started/tournament_results.rst]: I've added that in the wording below. Repeated it for both which might be a bit overkill, happy to tweak further :)

File: axelrod/result_set.py (hunk 3 above)
Comment (drvinceknight) [inline axelrod/result_set.py]: Added this and a bunch of others that were missing (whilst doing this I got annoyed at my prior lazy self).

File: axelrod/result_set.py (hunk 1 above)
Comment (drvinceknight) [inline axelrod/result_set.py]: 👍

File: axelrod/result_set.py (hunk 2 above)
Comment (drvinceknight) [inline axelrod/result_set.py]: Looks like it wasn't needed. Have removed it.

File: axelrod/result_set.py (hunk 7 above)
Comment (drvinceknight) [inline axelrod/result_set.py]: What did you mean by an integer division error? (I might have misunderstood.)

Comment (drvinceknight) [inline axelrod/result_set.py]: I know what you mean. Am pushing a better hypothesis test and a fix.
